runtime-transaction/src/compute_budget_program_id_filter.rs

```diff
@@ -1,37 +1,114 @@
 // static account keys has max
 use {
     agave_transaction_view::static_account_keys_frame::MAX_STATIC_ACCOUNTS_PER_PACKET as FILTER_SIZE,
-    solana_builtins_default_costs::MAYBE_BUILTIN_KEY, solana_sdk::pubkey::Pubkey,
+    solana_builtins_default_costs::{BUILTIN_INSTRUCTION_COSTS, MAYBE_BUILTIN_KEY},
+    solana_sdk::pubkey::Pubkey,
 };
 
+#[derive(Default, PartialEq)]
+enum BuiltinCheckStatus {
+    #[default]
+    Unchecked,
+    NotBuiltin,
+    Builtin {
+        is_compute_budget: bool,
+    },
+}
+
 pub(crate) struct ComputeBudgetProgramIdFilter {
-    // array of slots for all possible static and sanitized program_id_index,
-    // each slot indicates if a program_id_index has not been checked (eg, None),
-    // or already checked with result (eg, Some(result)) that can be reused.
-    flags: [Option<bool>; FILTER_SIZE as usize],
+    auxiliary_data: [BuiltinCheckStatus; FILTER_SIZE as usize],
 }
 
 impl ComputeBudgetProgramIdFilter {
     pub(crate) fn new() -> Self {
         ComputeBudgetProgramIdFilter {
-            flags: [None; FILTER_SIZE as usize],
+            auxiliary_data: core::array::from_fn(|_| BuiltinCheckStatus::default()),
         }
     }
 
     #[inline]
-    pub(crate) fn is_compute_budget_program(&mut self, index: usize, program_id: &Pubkey) -> bool {
-        *self
-            .flags
+    // Get or check instruction's Builtin auxiliary data,
+    // If instruction at `index` is a builtin, returns Some<is_compute_budget_instruction>
+    //   returns None if not a builtin instruction.
+    pub(crate) fn get_auxiliary_data(&mut self, index: usize, program_id: &Pubkey) -> Option<bool> {
+        let stat = self
+            .auxiliary_data
             .get_mut(index)
-            .expect("program id index is sanitized")
-            .get_or_insert_with(|| Self::check_program_id(program_id))
+            .expect("program id index is sanitized");
+        if stat == &BuiltinCheckStatus::Unchecked {
+            *stat = Self::check_status(program_id)
+        }
+
+        match stat {
+            BuiltinCheckStatus::NotBuiltin => None,
+            BuiltinCheckStatus::Builtin { is_compute_budget } => Some(*is_compute_budget),
+            _ => unreachable!("already checked"),
+        }
     }
 
     #[inline]
-    fn check_program_id(program_id: &Pubkey) -> bool {
+    fn check_status(program_id: &Pubkey) -> BuiltinCheckStatus {
         if !MAYBE_BUILTIN_KEY[program_id.as_ref()[0] as usize] {
-            return false;
+            return BuiltinCheckStatus::NotBuiltin;
         }
-        solana_sdk::compute_budget::check_id(program_id)
+
+        BUILTIN_INSTRUCTION_COSTS
```

Comment:
This is where we should be checking if builtins have been converted to sBPF programs, right? We should be checking if the following features are activated: `migrate_stake_program_to_core_bpf`, `migrate_config_program_to_core_bpf`, and `migrate_address_lookup_table_program_to_core_bpf`.

Maybe the type of `BUILTIN_INSTRUCTION_COSTS` can be changed to `AHashMap<Pubkey, BuiltinCost>` where `BuiltinCost` is:

```rust
struct BuiltinCost {
  native_cost: u64,
  sbpf_migration_feature: Option<Pubkey>,
}
```

and then if we get a builtin cost with `sbpf_migration_feature.is_some()` we can check if it's activated already.

Reply:
Great point. each builtin migration is feature gated individually, can't simply remove migrated builtin from the static map! Your suggestion works, perhaps it should be a separate prerequisite PR to this one? With some note that `BUILTIN_INSTRUCTION_COSTS` should be changed whenever builtin migration feature gate is added?

Reply:
Yeah that works for me

Reply:
Ok, working on it.

Reply:
#3768